### ISSUE TITLE
mnist.py: failed to allocate 8.00G from device: hipError_t(1002)

### ISSUE THREAD

**Issue by yige-hu** [ROCmSoftwarePlatform/hiptensorflow#62]
Operating System: Ubuntu 16.04.3 LTS, Linux kernel 4.13.0
GPU: AMD RX 580, 8GB memory
ROCm backend.

I installed hipTensorflow from the source, following the steps in:
https://github.com/ROCmSoftwarePlatform/hiptensorflow/blob/hip/rocm_docs/tensorflow-build-from-source.md
The commit hash of hipTensorflow: 28243d4
Bazel version: 0.4.5

Then, I tried to run the mnist.py in hiptensorflow examples:
https://github.com/ROCmSoftwarePlatform/hiptensorflow/blob/hip/tensorflow/examples/learn/mnist.py
And got the followed error message:
> $ python mnist.py
> I tensorflow/stream_executor/dso_loader.cc:139] successfully opened CUDA library libhipblas.so locally
> I tensorflow/stream_executor/dso_loader.cc:139] successfully opened CUDA library libhipfft.so locally
> I tensorflow/stream_executor/dso_loader.cc:139] successfully opened CUDA library libhip_hcc.so locally
> I tensorflow/stream_executor/dso_loader.cc:139] successfully opened CUDA library libhiprng.so locally
> I tensorflow/stream_executor/dso_loader.cc:139] successfully opened CUDA library libMIOpen.so locally
> Extracting MNIST-data/train-images-idx3-ubyte.gz
> Extracting MNIST-data/train-labels-idx1-ubyte.gz
> Extracting MNIST-data/t10k-images-idx3-ubyte.gz
> Extracting MNIST-data/t10k-labels-idx1-ubyte.gz
> WARNING:tensorflow:Using temporary folder as model directory: /tmp/tmpJJjENP
> WARNING:tensorflow:From mnist.py:94: calling fit (from tensorflow.contrib.learn.python.learn.estimators.estimator) with y is deprecated and will be removed after 2016-12-01.
> Instructions for updating:
> Estimator is decoupled from Scikit Learn interface by moving into
> separate class SKCompat. Arguments x, y and batch_size are only
> available in the SKCompat class, Estimator will only accept input_fn.
> Example conversion:
>   est = Estimator(...) -> est = SKCompat(Estimator(...))
> WARNING:tensorflow:From mnist.py:94: calling fit (from tensorflow.contrib.learn.python.learn.estimators.estimator) with x is deprecated and will be removed after 2016-12-01.
> Instructions for updating:
> Estimator is decoupled from Scikit Learn interface by moving into
> separate class SKCompat. Arguments x, y and batch_size are only
> available in the SKCompat class, Estimator will only accept input_fn.
> Example conversion:
>   est = Estimator(...) -> est = SKCompat(Estimator(...))
> WARNING:tensorflow:From mnist.py:94: calling fit (from tensorflow.contrib.learn.python.learn.estimators.estimator) with batch_size is deprecated and will be removed after 2016-12-01.
> Instructions for updating:
> Estimator is decoupled from Scikit Learn interface by moving into
> separate class SKCompat. Arguments x, y and batch_size are only
> available in the SKCompat class, Estimator will only accept input_fn.
> Example conversion:
>   est = Estimator(...) -> est = SKCompat(Estimator(...))
> WARNING:tensorflow:From /usr/local/lib/python2.7/dist-packages/tensorflow/contrib/learn/python/learn/estimators/head.py:1362: scalar_summary (from tensorflow.python.ops.logging_ops) is deprecated and will be removed after 2016-11-30.
> Instructions for updating:
> Please switch to tf.summary.scalar. Note that tf.summary.scalar uses the node name instead of the tag. This means that TensorFlow will automatically de-duplicate summary names based on the scope they are created in. Also, passing a tensor or list of tags to a scalar summary op is no longer supported.
> W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE3 instructions, but these are available on your machine and could speed up CPU computations.
> W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
> W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
> W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
> W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
> W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
> W tensorflow/stream_executor/cuda/cuda_driver.cc:633] creating context when one is currently active; existing: 0x7f0ec9177080
> I tensorflow/core/common_runtime/gpu/gpu_device.cc:892] Found device 0 with properties: 
> name: Device 67df
> major: 3 minor: 0 memoryClockRate (GHz) 1.36
> pciBusID 0000:01:00.0
> Total memory: 8.00GiB
> Free memory: 7.75GiB
> I tensorflow/core/common_runtime/gpu/gpu_device.cc:913] DMA: 0 
> I tensorflow/core/common_runtime/gpu/gpu_device.cc:923] 0:   Y 
> I tensorflow/core/common_runtime/gpu/gpu_device.cc:972] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Device 67df, pci bus id: 0000:01:00.0)
> E tensorflow/stream_executor/cuda/cuda_driver.cc:1056] failed to allocate 8.00G (8589934592 bytes) from device: hipError_t(1002)
> Illegal instruction (core dumped)

Is there anything I can do (e.g. change settings in hiptensorflow) to resolve this? Or can I decrease the input size by tuning parameters of mnist.py?

I really appreciate your help!

**Comment by whchung** [comment]
@yige-hu could you try reduce batch size?

**Comment by yige-hu** [comment]
@whchung 
For hiptensorflow/tensorflow/examples/learn/mnist.py, even after reducing the batch_size=1 in the script, I'm still getting the error message. I doubt if it's an hipTF implementation error, since it's allocating a memory buffer size equals to the total memory (8 GB), which is greater than the free memory (7.75 GB):
> Total memory: 8.00GiB
Free memory: 7.75GiB

With the mnist.py from https://github.com/tensorflow/models.git (tutorials/image/mnist), I'm getting an
> Illegal instruction (core dumped)

Similar to #61 , the instruction cannot be shown in asm layout from the core dumped.

Thanks,
Yige

**Comment by evg-tyurin** [comment]
I had similar warning on RX480. 

At first, 
Total memory: 8.00GiB
Free memory: 7.75GiB

Then, 
failed to allocate 7.36G 

Then script continued and Tensorflow successfully recreated the device. After few experiments desktop stop responding and I've restarted the computer. After restart, all is OK without any warnings.

**Comment by Sumenia** [comment]
same here with AMD Vega 56

**Comment by yige-hu** [comment]
BTW, I have a patch to fix this. @whchung Could you take a look at pull request #68

**Comment by whchung** [comment]
Please kindly try https://github.com/ROCmSoftwarePlatform/tensorflow/ and open the issue there in case it persists. This project has been deprecated and I'll close this repository very soon. Thanks for your understanding.
